### PR TITLE
Implement a driver, HIL, and capsule to write to flash

### PR DESCRIPTION
**Pull request by ia0**
### Pull Request Overview

The purpose of the capsule is to provide low-level control of the embedded flash to allow applications to implement flash-efficient data-structures using their writeable flash regions. The API is blocking since most flash either halt the CPU during write and erase operations or ask the application to wait until the operation is finished. A blocking API is also simpler to reason and less error-prone.

### Testing Strategy

In addition to `make ci-travis`, this pull request was tested on the `nrf52840dk` board by implementing in `libtock-rs`:
- a syscall interface
- a multi-set data-structure
- an example application similar to `blink.rs` but with a persistent counter

And updating the linker script.

### TODO or Help Wanted

This pull request may use feedback for other boards if needed, although this may as well be done in followup pull-requests.

### Documentation Updated

- [x] Updated the relevant files in `/docs`.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by bradjc** [comment]
Overall this looks like a nice set of changes. But how does this compare to the [flash hil](https://github.com/tock/tock/blob/master/kernel/src/hil/flash.rs), [nonvolatile storage hil](https://github.com/tock/tock/blob/master/kernel/src/hil/nonvolatile_storage.rs), and [app flash driver](https://github.com/tock/tock/blob/master/capsules/src/app_flash_driver.rs) that we already have?

**Comment by ia0** [comment]
Good question. TLDR: The main differences are the low-level control of the flash (enabling the implementation of flash-efficient and consistent data-structures in user-land) and the blocking API (not very important but a small plus for a safety oriented kernel).

### Related APIs and drivers

#### kernel/src/hil/flash.rs

The API is:
- Read page.
- Write page.
- Erase page.

Trying to implement a flash-efficient data-structure that can recover (completing or rolling-back to reach a consistent state) from partial operations (e.g. when power is lost during a write or erase operation) would not be performant in terms of flash and CPU. For example, flipping a bit in a word would require to write the whole page.

Also the API is non-blocking which adds unnecessary error-prone administration.

#### kernel/src/hil/nonvolatile_storage.rs

The API is:
- Read slice.
- Write slice.

This is not possible to erase a page with this API.

Also the API is non-blocking which adds unnecessary error-prone administration.

#### capsules/src/app_flash_driver.rs

The API is:
- Write slice.

This is not possible to erase a page with this API.

Also the API is non-blocking which adds unnecessary error-prone administration.

Also this capsule allows applications to write all their flash (including `.text` etc.). Although this may be useful for some applications, it feels safer to only allow write access to the writeable flash regions.

**Comment by bradjc** [comment]
The flash HIL works on pages because flash typically requires it. The nRF flash is a little weird, but still has "flash-like" restrictions:


> Section 4.3 NVMC:
>As illustrated in Memory on page 20, the flash is divided into multiple pages. The same 32-bit word in the flash can only be written n WRITE number of times before a page erase must be performed.
>
> The NVMC is only able to write 0 to bits in the flash that are erased (set to 1). It cannot rewrite a bit back to 1. Only full 32-bit words can be written to flash using the NVMC interface. To write less than 32 bits, write the data as a full 32-bit word and set all the bits that should remain unchanged in the word to 1. Note that the restriction on the number of writes (nWRITE) still applies in this case.

Looking at `nWRITE` it is 2, so it is not reliable to expect that writing just a word on the nRF will work for general purpose applications.

The nonvolatile storage HIL is designed for persistent storage with non "flash-like" properties (e.g. FRAM).

#### Block v. non-blocking

I think for better or worse low-level operations are split-phase, and we have to deal with the complexity that brings. Tock's design has been to disallow blocking in the kernel, and to provide abstractions in userspace for writing synchronous looking code.

#### Userspace driver

Having a new driver that enables boards to restrict apps to only modifying their writeable flash regions seems like a useful addition.

**Comment by ia0** [comment]
> The flash HIL works on pages because flash typically requires it.

What do you mean? Some flashes can only write a whole page? If that's the case, then they simply need to define their word size to be the page size. In the proposed HIL, the word size is the smallest unit of flash that can be written. In nRF52840dk it happens to be a `u32` and called a word, so I reused that terminology.

> Looking at `nWRITE` it is 2, so it is not reliable to expect that writing just a word on the nRF will work for general purpose applications.

This is not expected and is actually documented in the `Safety` section of `write_slice` in `kernel/src/hil/embedded_flash.rs`. I also used to document the fact that the flash only writes 1s to 0s, but removed it because I thought it was common knowledge. I can add it back if some flashes are different. And maybe I should document all that in the `doc/syscalls/50003_embedded_flash.md`? Please advise.

The most useful input I could get are a good description of what kind of flashes we find out there (since I only looked at nRF52). Are any of them not satisfying the following properties?
- They can only switch 1s to 0s (leaving some 1s as they are) and that only at a certain granularity (that I call word in the API since for nRF52 it happens to be a word).
- They can only switch all 0s to 1s by erasing a full page.
- They can only write a word a limited amount of time between page erasures.
- They can only erase a page a limited amount of time during the lifetime of the flash.
- They forbid using the flash while an operation is ongoing (some of them like the nRF52 actually halts the CPU).

The HIL was designed with those kind of flashes in mind. I'm interested to learn if there are others, and if Tock currently supports them.

> I think for better or worse low-level operations are split-phase, and we have to deal with the complexity that brings. Tock's design has been to disallow blocking in the kernel, and to provide abstractions in userspace for writing synchronous looking code.

I think this is a matter of compromise. If flashes (like nRF52) halt the CPU during flash operations, it's essentially blocking everything (kernel and applications), so a non-blocking interface is only adding complexity. But if we know some flashes still permit to use the flash or do other things during an operation, I guess we could add a non-blocking API but it would be essentially extending the unsafe surface since the application has to take care of not violating those restrictions. What do you think?

[FYI: I'll be offline from Saturday to Tuesday.]

**Comment by bradjc** [comment]
I'm not sure that documenting the issues around writing using this interface is enough in this case. I think it is "anti-Tock" to expose all the way to userspace an interface that doesn't really work. An app can call write to its app flash region, but it might not actually store what it tried to write, and I don't think it should be up to the app to keep track of those details.

The capsule also looks to be unsafe right now, because apps I think have the ability to call erase on a page which might contain data not owned by the app.

We are working on better documenting things like this, but a design decision of the Tock kernel is that it does not contain blocking interfaces. So any flash HIL will have to have a callback when the operation is done.

**Comment by ia0** [comment]
> I'm not sure that documenting the issues around writing using this interface is enough in this case. I think it is "anti-Tock" to expose all the way to userspace an interface that doesn't really work. An app can call write to its app flash region, but it might not actually store what it tried to write, and I don't think it should be up to the app to keep track of those details.

Yes, this is a low-level API so it exposes some unsafe parts (that making safe would require using a shadow memory and thus use kernel RAM). If Tock doesn't want to expose such APIs to userland, it means it is not possible to write efficient (regarding CPU, RAM, and flash usage), persistent, consistent (can recover from power failures), flash-friendly (does not erase pages too much) data-structures in userland. This means the kernel has to provide those data-structures. Which means the kernel gets more complicated and operating with the data-structures involves many syscalls. I don't think such alternative is better for Tock and makes it safer, on the contrary. The question here is to find the right balance between exposing "unsafe" functions to userland and a big kernel with slower drivers.

> The capsule also looks to be unsafe right now, because apps I think have the ability to call erase on a page which might contain data not owned by the app.

How? The capsule checks line 89 that the page is in some application writeable flash region.

> We are working on better documenting things like this, but a design decision of the Tock kernel is that it does not contain blocking interfaces. So any flash HIL will have to have a callback when the operation is done.

Ok, I'll make the API non-blocking with the usual Tock patterns. I don't think it's beneficial but it's not hurting much either.

**Comment by bradjc** [comment]
I'm not sure I agree that an API like this is useful, _but_ we do have low-level userspace APIs for other resources (e.g. SPI buses), so I can see the argument for having a low level API for flash as well. Given that:

- Can the write function just be added to the existing flash HIL? I think it is confusing to have two flash HILs.

- I think the capsule should either be expanded to support a more reliable app flash region driver in addition to the very low level interface, or the capsule should be renamed to leave the current name open for one that provides a more reliable interface. My guess is app authors would expect the more reliable interface.

As for the erase issue, what happens on a platform with 4096 byte pages if a hypothetical app has a writeable region like:

```
0x1FF0: start of app writeable region
0x7FFF: end of app writeable region
```
and the app calls erase with a pointer of `0x1FF0`? Won't all of the flash from `0x1000` to `0x1FFF` be erased?

**Comment by ia0** [comment]
> * Can the write function just be added to the existing flash HIL? I think it is confusing to have two flash HILs.

I agree it is confusing to have multiple related HILs. However, one advantage of having multiple HILs is that a driver does not need to implement all of them, but only those where it is possible, necessary, or useful. It provides more flexibility. The disadvantage is that we would need to document clearly which HILs has which purpose. Note that if we're fine returning `ENOSUPPORT` for some parts of the HIL, then we can combine both advantages by having only one HIL with possibly partial implementation. What do you think?

> * I think the capsule should either be expanded to support a more reliable app flash region driver in addition to the very low level interface, or the capsule should be renamed to leave the current name open for one that provides a more reliable interface. My guess is app authors would expect the more reliable interface.

I agree that application authors should first aim for a safe capsule. Since I don't know how a safe API should look like (I don't have such usage in mind), I would prefer renaming this capsule to strengthen the fact that it is low-level and should not be a first pick. What about `LowLevelEmbeddedFlash` or `UnsafeEmbeddedFlash`? Or would it make sense to only have these annotations at the command level with only one capsule?

> As for the erase issue, what happens on a platform with 4096 byte pages if a hypothetical app has a writeable region like:
> 
> ```
> 0x1FF0: start of app writeable region
> 0x7FFF: end of app writeable region
> ```
> 
> and the app calls erase with a pointer of `0x1FF0`? Won't all of the flash from `0x1000` to `0x1FFF` be erased?

The capsule would check if [0x1FF0, 0x2FF0) is in a writeable flash region and succeed. It will then call the driver that will check if the address is page-aligned line 432 and return `EINVAL` since it's not.

**Comment by bradjc** [comment]
For interfaces that we know only a subset of chips can implement, we have been doing "advanced" interfaces. For example with UART and different receive operations. I think using an error like unsupported is only good in cases where the implementation doesn't exist because no one has written it. With flash, however, I think it is reasonable to just expect each chip to implement a write slice operation.

For the name, since the capsule is still restricted to just the writeable app region it should reflect that as well.

**Comment by alevy** [comment]
tl;dr I think we can get from this PR to something we would definitely want to include, and i think we can do it without violating kernel design principles.

I’ll try to summarize the positives and concerns about this PR in broad strokes (i.e. not with regard to any specifics in the implementation, which is overall very good).

First, I agree with @bradjc that this is overall a well done PR, thank you @ia0! In terms of the quality of the code, if we all agreed on the general design, this would be an easy PR to merge with maybe only minor comments.

Second, if I’m interpreting the goals of this PR—-the high-level problem it is meant to address—-I think it’s goals that we would definitely like to address in some form.

Finally,  I also agree with @bradjc, that this particular PR violates a couple design principles the kernel follows.

Those design principles are not very well documented anywhere central yet (this is a point of discussion recently), so it’s not really fair to expect you (or any other newer contributor) to know them, so I’ll try to restate them here.

Of course, like any design principle, sometimes it can make sense to violate the principle, but if we choose to do so we have to be clear about it, and make sure it’s worth it.

# Some in-kernel design principles

## Role of HILs

Generally, the Tock kernel is structured into three layers:

  1. Chip-specific drivers: these typically live in a crate in the`chips` subdirectory, or an equivalent crate in an different repo (e.g. the Titan port is out of tree but it’s `h1b` create is the equivalent here). These drivers have implements that are specific to the hardware particular to a certain microcontroller. Ideally, their implementation is fairly simple, and they merely adhere to a common interface (a HIL). That’s not always the case, but that’s the ideal.

  2. Chip-agnostic, portable, peripheral drivers and subsystems. These typically live in the `capsules` crate. These includes things like the virtual alarms and virtual I2C stack, as well as drivers for hardware peripherals not on the chip itself (e.g. sensors, radios, etc). These drivers typically rely on the chip-specific drivers through the HILs.

  3. System call drivers, also typically found in the `capsules crate. These are the drivers that implement a particular part of the system call interfaces, and are often even more abstracted from the hardware than (2)—-for example, the temperature sensor system call driver can use _any_ temperature sensor, including several implemented as portable peripheral drivers. We don’t have many examples of this, but the system call interface is _another_ point of standardization that can be implemented in various ways. So it’s perfectly reasonable to have several implementations of the same system call interface that use completely different hardware stacks, and therefor HILs and chip-specific drivers (e.g. a console driver that operates over USB might just be implemented as a different system call driver that implements the same system calls, rather than trying to fit USB into the UART HIL).

The connective tissue between layers are the HILs. The HIL interfaces are portable interfaces that are implemented in a non-portable way.

The choice of particular HIL interfaces is pretty important, and we have some general principles we follow:

  1. HIL implementations get to assume this HIL is the _only_ way the device will be used. As a result, it’s generally a bad idea to have several HILs that provide different interfaces to similar functionality, because it will not, in general, be possible for multiple drivers to use different HILs for the same device simultaneously. Clearly the flash HIL and nonvolatile storage HIL are counter examples to this principle, but that’s not necessarily good (and it definitely shouldn’t be exacerbated if it’s not good).

  2. HIL implementations should be fairly general. If we have an interface that doesn’t work very well across different hardware, we probably have the wrong interface—-it’s either too high level, or too low level, or it’s just not flexible enough. But HILs shouldn’t generally be designed to optimize for _particular_ applications or hardware, and definitely not for a particular _combination_ of applications and hardware. If there are cases where that is really truly necessary, you can always implement a driver that is chip or board specific, and circumvents the HILs entirely.

## Split-phase operation

While processes are time sliced and preemptive in Tock, the kernel is _not_. Everything is run-to-completion. That’s an important design choice because it allows the kernel to avoid allocating lots of stacks for lots of tasks, and it makes it possible to reason more simply about static and other shared variables.

As a result, by design, all I/O operations _have_ to be asynchronous, so that the kernel can be reasonably counted on to operate in a timely manner. We do this using TinyOS style split-phase callbacks because they are resolved statically and it avoids the need to allocate closures.

Drivers are indeed more cumbersome to write than with a blocking API, however, this is a conscious choice to favor overall safety of the kernel (e.g. avoiding running out of memory or preventing other code from running on time) over functional correctness of individual drivers (because they might be more error-prone, not because they _can’t_ be written correctly).

There are cases where we violate this. For example, the SAM4L’s GPIO controller may take up to 5 cycles to become ready between operations. Technically, according to this principle, the GPIO driver should therefore be split-phase to wait on the ready condition, but we know it’ll take longer to set that up than to just spin on the ready bit, so we just spin for at most a handful of cycles. But those cases are rare.

# This PR

Because this PR goes against these two principles, it’s cause for some higher-level investigation. Is this actually necessary? Is there a better way of solving the high level problem without violating one or both of the principles?

I think there are two, independent, changes you are proposing related to flash.

## Finer grained flash operations

You want more fine-gained access to flash from higher-level parts of the kernel.

I’m guessing that you want this in order to implement something like the knockdown counter in the Titan port—-you want a monotonically increasing counter that is resilient to power failures.

That seems super reasonable.

What changes to the flash HIL would you need to implement _specifically_ that? I’m almost certain you don’t need blocking I/O at the HIL level. I think you need to be able to make changes to the flash at a finer granularity than a whole page (I confess I never _really_ understood the knockdown algorithm, so I don’t really know exactly what you need).

## Blocking I/O

I think you actually raise an interesting point. In practice, flash writes/erased may _not_ be asynchronous anyway. This is of course _not_ true for flash storage not on the chip (e.g. an SD card or flash chip), but it might be generally true for the executable flash on most microcontrollers.

If that’s indeed the case, they it _could_ make sense to make an exception in the case of the flash HIL and just use block calls, since the resulting execution would be blocked regardless.

Conversely, if this is _only_ true for one or a few chips, I think those ports should just bite the bullet and deal with the complexity. Can we maybe survey a few representative chips and figure out how they treat flash? The ones we have support for make sense to start from.

# My Suggestions

My hope is that we can extend or change the flash HIL to support the necessary features for use cases such as a secure monotonic counter (or whatever other data structures). I also hope that we can either agree that all flash modifying operations should just be blocking, or that _no_ flash modifying operations should be blocking.

The result would be still a single flash HIL, not a new one.

I suggest we start by proposing functional changes at the HIL level (i.e. finer grained erase/writes). Then, separately, discuss blocking write/erases for the flash.

I also suggest that an issue would be a better place to start this.

**Comment by ia0** [comment]
Thanks for your explanation! I created #1459.

Do you have an example of "a driver that is chip or board specific, and circumvents the HILs entirely"? (in case it happens that the flash of nrf52[840dk] is very different from other flashs)

**Comment by ia0** [comment]
If everyone agrees, this PR can probably be closed. A chip-specific driver is being implemented in #1467.

**Comment by bradjc** [comment]
Moved to #1467
